geomesa-hbase/geomesa-hbase-datastore/src/main/scala/org/locationtech/geomesa/hbase/data/HBaseQueryPlan.scala

```diff
@@ -66,11 +70,34 @@ case class ScanPlan(filter: HBaseFilterStrategyType,
                     resultsToFeatures: Iterator[Result] => Iterator[SimpleFeature]) extends HBaseQueryPlan {
   override def scan(ds: HBaseDataStore): CloseableIterator[SimpleFeature] = {
     ranges.foreach(ds.applySecurity)
-    val results = new HBaseBatchScan(ds.connection, table, ranges, ds.config.queryThreads, 100000, remoteFilters)
+    val results: HBaseBatchScan = new HBaseBatchScan(ds.connection, table, ranges, ds.config.queryThreads, 100000, remoteFilters)
     SelfClosingIterator(resultsToFeatures(results), results.close)
   }
 }
 
+case class CoprocessorPlan(sft: SimpleFeatureType,
+                           filter: HBaseFilterStrategyType,
+                           hints: Hints,
+                           table: TableName,
+                           ranges: Seq[Scan],
+                           remoteFilters: Seq[HBaseFilter] = Nil,
+                           resultsToFeatures: Iterator[Result] => Iterator[SimpleFeature]) extends HBaseQueryPlan  {
+  /**
+    * Runs the query plain against the underlying database, returning the raw entries
+    *
+    * @param ds data store - provides connection object and metadata
+    * @return
+    */
+  override def scan(ds: HBaseDataStore): CloseableIterator[SimpleFeature] = {
+    val is: Map[String, String] = KryoLazyDensityCoprocessor.configure(sft, null, hints)
```

Comment:
right now there is only the one coprocessor, but ti would still be good to add a check here something like
```
if (hints.isDensityQuery) {
  ...
} else {
  throw new NotImplementedExcecption
}
```